include/rocksdb/utilities/object_registry.h

```diff
@@ -249,13 +255,136 @@ class ObjectRegistry {
     }
   }
 
+  // Sets the object for the given id/type to be the input object
+  // If the registry does not contain this id/type, the object is added and OK
+  // is returned. If the registry contains a different object, an error is
+  // returned. If the registry contains the input object, OK is returned.
+  template <typename T>
+  Status SetManagedObject(const std::shared_ptr<T>& object) {
+    assert(object != nullptr);
+    return SetManagedObject(object->GetId(), object);
+  }
+
+  template <typename T>
+  Status SetManagedObject(const std::string& id,
+                          const std::shared_ptr<T>& object) {
+    const auto c = std::static_pointer_cast<Customizable>(object);
+    return SetManagedObject(T::Type(), id, c);
+  }
+
+  // Returns the object for the given id, if one exists.
+  // If the object is not found in the registry, a nullptr is returned
+  template <typename T>
+  std::shared_ptr<T> GetManagedObject(const std::string& id) const {
+    auto c = GetManagedObject(T::Type(), id);
+    return std::static_pointer_cast<T>(c);
+  }
+
+  // Returns the set of managed objects found in the registry matching
+  // the input type and ID.
+  // If the input id is not empty, then only objects of that class
+  // (IsInstanceOf(id)) will be returned (for example, only return LRUCache
+  // objects) If the input id is empty, then all objects of that type (all Cache
+  // objects)
+  template <typename T>
+  Status ListManagedObjects(const std::string& id,
+                            std::vector<std::shared_ptr<T>>* results) const {
+    std::vector<std::shared_ptr<Customizable>> customizables;
+    results->clear();
+    Status s = ListManagedObjects(T::Type(), id, &customizables);
+    if (s.ok()) {
+      for (const auto& c : customizables) {
+        results->push_back(std::static_pointer_cast<T>(c));
+      }
+    }
+    return s;
+  }
+
+  template <typename T>
+  Status ListManagedObjects(std::vector<std::shared_ptr<T>>* results) const {
+    return ListManagedObjects("", results);
+  }
+
+  // Creates a new ManagedObject in the registry for the id if one does not
+  // currently exist.  If an object with that ID already exists, the current
+  // object is returned.
+  //
+  // The ID is the identifier of the object to be returned/created and returned
+  // in result
+  // If a new object is created (using the object factories), the cfunc
+  // parameter will be invoked to configure the new object.
+  template <typename T>
+  Status GetOrCreateManagedObject(const std::string& id,
```

Comment:
One general question, the managed object can only be shared within the same process right? For example, a process create N DB instance and one Cache instance, so the cache is shard by N DB. So once the ID is the same, we always get the ptr of the same object instance. LoadSharedObject ensure that, each time we call it, it will return different instance even we give the same ID?

Reply:
@zhichao-cao There is only sharing of objects within a process.  There is nothing (using shared memory, etc) to allow multiple processes on a single host.  Is there a requirement/request to do so?  I do not know of any objects that are currently shared that way now.   It would be theoretically possible for a MemoryAllocator to be "shared" between processes, where each process had their own Allocator referencing a common/shared pool.  This would be a feature of how the allocator was set up and I do not know how locks/concurrency would be addressed between the processes

Reply:
For now, without the ManagedObject related interface, the way we do object sharing across DB is, the main thread create this object and pass the shared_ptr to different DB instances. For now, with this change, the main thread only need to remember the ID? The object is always alive until the process is ended since it is managed by the objectregestry?

Reply:
> For now, without the ManagedObject related interface, the way we do object sharing across DB is, the main thread create this object and pass the shared_ptr to different DB instances. For now, with this change, the main thread only need to remember the ID? The object is always alive until the process is ended since it is managed by the objectregestry?

@zhichao-cao If someone does not maintain a reference to the underlying object, then it will be removed from the ObjectRegistry (it is stored as a weak pointer).  For purposes here, I will assume the object being shared is Cache, but it applies to any managed object.  There are a couple of ways this can be handled.
1.  The existing way.  The main thread creates the cache and then passes the ID (or shared ptr) to the other DBs.  Each DB can get the cache from its ID (or shared ptr)
2. The OPTIONS file.  If the Cache is stored in the Options file and all DBs in the same process use the same cache, then all of them will have the same ID.  When the options are hydrated from disk, the ID will be read and all of the DBs in the same process will have stored the same ID in their options file, causing all of them to use the same managed object.

Note that the two methods cannot be combined.  If a cache is create from scratch in the main thread, it will have a different ID than the one in the options file and therefore will use a different cache instance.  If the cache is re-hydrated from the options file, then a new cache will be created and then destroyed/replaced by replacing the one passed in with the one re-created.